### PR TITLE
[FLINK-5418] [table] Estimated row size does not support nested types

### PR DESCRIPTION
**Pull request by twalthr**
This adds an estimate for array, composite and generic types.

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/FlinkRel.scala [hunk 1]
```diff
@@ -120,6 +120,7 @@ trait FlinkRel {
         case typeName if SqlTypeName.YEAR_INTERVAL_TYPES.contains(typeName) => s + 8
         case typeName if SqlTypeName.DAY_INTERVAL_TYPES.contains(typeName) => s + 4
         case SqlTypeName.TIME | SqlTypeName.TIMESTAMP | SqlTypeName.DATE => s + 12
+        case SqlTypeName.ARRAY | SqlTypeName.ROW | SqlTypeName.ANY => s + 24
```
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/FlinkRel.scala]: Any chance we can expand the ARRAY or the ROW type, and only counting the basic data structures to make this a little more accurate?

**Comment by fhueske** [comment]
A better estimate for `ROW` will be merged with #3039. I think the best we can do for array is to estimate the size of one array element and have some default array size.

**Comment by fhueske** [comment]
Hi @twalthr, I have merged PR #3039 yesterday. Can you update this PR? Thanks

**Comment by twalthr** [comment]
I will update this PR with the code from #3039 as a fix for the 1.2 release branch. What do you think is a good estimate for an array size? 16 elements? How do we want to estimate the size of a generic type?

**Comment by fhueske** [comment]
Good question, 16 sounds OK to me. For generic types I don't know, maybe 32 bytes?

**Comment by twalthr** [comment]
I have chosen 128 now. Because generic types should be very expensive. If they have 4-8 fields with arrays and strings we should already be in that region.

**Comment by fhueske** [comment]
+1 to merge

**Comment by twalthr** [comment]
Merging this...

**Pull request closed without merging**